### DISCUSSION TITLE
Registry namespaces

### DISCUSSION THREAD

**Discussion by lann** [bytecodealliance/SIG-Registries#8] [General]
Namespaces are terrible. No namespaces is terrible. We should start a discussion about whether namespaces should be allowed/required/forbidden for BCA-operated registries.

**Comment by lann** [comment]
I am moderately pro-namespaces. A pre-proposal:

* A BCA-operated/endorsed public registry should require namespaces, either of:
  * a registered second-level domain, e.g. `bytecodealliance.org`
    * validated automatically via conventional DNS (`TXT` record) or HTTP ([`.well-known/...`](https://datatracker.ietf.org/doc/html/rfc5785)) means
  * a "sub-namespace" with different authentication mechanism(s), e.g. `github.com/bytecodealliance`
    * the "parent" namespace would be manually configured by the registry operator

**Reply by michelleN** [reply]
@lann Just so I'm on the same page, namespaces are used to group an org's packages/components/entities?

I really like your proposal for how a BCA operated/endorsed registry should require namespaces. This helps somewhat prevent the squatting issue and also relate an actual org to a namespace. good stuff.

In this this scenario, if there is an org called example has registered the domain example.com and another org which has the registered domain example.org, only one of them would be able to claim the namespace `example` right? and thus also only one of them would be able to register that namespace in the public registry via the DNS validation procedure? Am I understanding that correctly?

**Reply by lann** [reply]
I think the namespace would include the TLD, so every namespace would end in e.g. `.com`, `.org`; `example.com`/`example.org` would be separate namespaces.

**Reply by carlsverre** [reply]
I am a fan of requiring namespaces for formal registries but keeping them optional in the API layer. I am less sure about requiring that namespaces are attached to domain names. Consider a small open source community with a discord server. Should they require attaching the registry to a domain name?

**Reply by lann** [reply]
> Consider a small open source community with a discord server. Should they require attaching the registry to a domain name?

I would hope that a "sub-namespace" would meet their needs, e.g. `gitlab.com/small-community`. The trouble with _not_ tying the namespace to a domain is namespace squatting and related administrative/legal overhead.

**Reply by guybedford** [reply]
Would domain names be used as a primary means of authentication, or only at the point of initial registration as a form of Sybil resistance mechanism?

I quite like the idea of them being used initially, then no longer required - because domain names expire and people often forget to renew credit card details. Especially if the domain name requirement means one might be incentivized to reach for some random domain to have to hand just to get going with a publishing workflow. See for example the npm security discussion of the day if you haven't already (https://twitter.com/lrvick/status/1523774962909298690?s=20&t=HSAeYazCzANerbVZ9fk4Gg).

Nice to close the door on any of those types of attacks by making the established authenticated the primary authority once setup, without the TLD having any residual authority in the system.

It's always an option to make namespaces optional and primarily for orgs ala npm to retain low requirements to get started for individual developers - best and worst of both worlds :)

**Reply by lann** [reply]
> Would domain names be used as a primary means of authentication, or only at the point of initial registration as a form of Sybil resistance mechanism?

I hadn't thought through this specifically until today. I think it makes sense that "proof of ownership" would only be needed to _bootstrap_ authorization for a namespace in most circumstances, though it brings up a question of what happens to abandoned or deleted namespaces, which is the subject of https://github.com/bytecodealliance/SIG-Registries/issues/12. My suggestion in that issue is to allow transfer of ownership of a namespace/package, but treat it as a totally new package for update purposes, preventing any kind of automated updates across that "ownership boundary", which would have helped in the npm case.
